### PR TITLE
Replace SAO with ASSAO as Godot's new SSAO

### PR DESCRIPTION
**Pull request by clayjohn**
Supersedes: https://github.com/godotengine/godot/pull/42077

This is a direct port of Intel's [ASSAO](https://github.com/GameTechDev/ASSAO) using some optimizations from AMD's [CACAO](https://github.com/GPUOpen-Effects/FidelityFX-CACAO).

This is significantly faster and higher quality than our current SSAO. It is also much higher quality than MSSAO in #42077 but is slower. 

Timing on RX 5600 XT
ASSAO
|                   | Lowest Quality | Highest Quality|
|-------------|-----------------|------------------|
| Full Size     | 0.2 ms     |  1.1 ms    |
| Half Size    |  0.12 ms    |   0.46 ms        |

SAO (currently broken on master)
|                   | Lowest Quality | Highest Quality|
|-------------|-----------------|------------------|
| Full Size     | 2 ms    |  4ms   |
| Half Size    |      |           |

Old SAO
sao-hq-sponza

New ASSAO
ASSAO-uq-sponza

New ASSAO - Low Quality - Half Resolution
ASSAO-lq-hr-sponza

<details><summary>Not used MSSAO</summary>

mssao-hq-sponza

</details>

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/ssao_importance_map.glsl [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+	// sum the average; to avoid overflowing we assume max AO resolution is not bigger than 16384x16384; so quarter res (used here) will be 4096x4096, which leaves us with 8 bits per pixel
+	uint sum = uint(clamp(retVal, 0.0, 1.0) * 255.0 + 0.5);
+
+	// save every 9th to avoid InterlockedAdd congestion - since we're blurring, this is good enough; compensated by multiplying LoadCounterAvgDiv by 9
+	if (((ssC.x % 3) + (ssC.y % 3)) == 0) {
+		atomicAdd(counter.sum, sum);
+	}
+#endif
+}
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/shaders/ssao_importance_map.glsl:133]: Missing newline at EOF.

File: servers/rendering/renderer_rd/shaders/ssao_interleave.glsl [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	float a = textureLod(source_texture, vec3(uv, 0), 0.0).x;
+	float b = textureLod(source_texture, vec3(uv, 1), 0.0).x;
+	float c = textureLod(source_texture, vec3(uv, 2), 0.0).x;
+	float d = textureLod(source_texture, vec3(uv, 3), 0.0).x;
+	float avg = (a + b + c + d) * 0.25;
+
+#endif
+	imageStore(dest_image, ivec2(gl_GlobalInvocationID.xy), vec4(avg));
+#endif
+}
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/shaders/ssao_interleave.glsl:119]: Missing newline at EOF.

**Comment by Calinou** [comment]
Is it intended that I can see the textures' details on the ASSAO screenshot? I can't on the SAO and MSSAO screenshots.

**Comment by clayjohn** [comment]
@Calinou yep! ASSAO uses the normal buffer to generate SSAO while our implementation of SAO ignored it and MSSAO couldn't use it. 

So what your seeing is not the texture's AO, but a high detail AO generated from normals.

**Comment by Calinou** [comment]
@clayjohn Does this mean AO maps are still relevant? What about https://github.com/godotengine/godot/issues/43957 ?

**Comment by clayjohn** [comment]
@Calinou AO texture maps are much higher quality than SSAO, plus they can be used even when the object doesn't have a normal map. There is no reason not to use both AO maps and SSAO together IMO.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
Look OK, I am still in doubt about exporting so many parameters to Environment, but I need to try them for a while to see what is actually important.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Zireael07** [comment]
Compute shaders again or something backportable to 3.x?

**Comment by Calinou** [comment]
> Compute shaders again or something backportable to 3.x?

Indeed, compute shaders are still required for this feature.

**Comment by clayjohn** [comment]
@Zireael07 @Calinou in theory the original ASSAO could be ported to GLES3 as it's all done in a pixel shader. However, it would require duplicating all the work I just did here.

**Comment by atirut-w** [comment]
> it would require duplicating all the work I just did here.

I think it would still be worth it to backport, though. IMO, there might be 3.x games that want realistic graphics but cannot move to 4.x for some reasons, and #48503 affects them.

**Comment by Calinou** [comment]
@atirut-w Someone was working on implementing GTAO in `3.x` instead: https://github.com/godotengine/godot-proposals/issues/3223

**Pull request merged**